### PR TITLE
Add infeasibility analysis tool

### PR DESCRIPTION
**Pull request by flomnes**

### CODE REVIEW

File: src/solver/infeasible-diag/problem.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+MPSolver::ResultStatus InfeasibleProblemDiag::Solve()
+{
+    return mSolver->Solve();
+}
+
+InfeasibleProblemReport InfeasibleProblemDiag::produceReport()
+{
+    addSlackVariables();
+    buildObjective();
+    assert(mSlackVariables.size() > 0);
```
Comment (flomnes) [inline src/solver/infeasible-diag/problem.cpp]: Absence of slack variables shouldn't be fatal : we only provide a analysis

File: src/solver/cmake/infeasible-diag.cmake [hunk 2]
```diff
@@ -0,0 +1,12 @@
+set(SRC_INFEASIBLE_DIAG
+  infeasible-diag/problem.cpp
+  infeasible-diag/problem.h
+  infeasible-diag/report.h
+  infeasible-diag/report.cpp
+  )
+
+add_library(libinfeasible_diag ${SRC_INFEASIBLE_DIAG})
```
Comment (flomnes) [inline src/solver/cmake/infeasible-diag.cmake]: Find a better name ?

**Review by flomnes — COMMENTED** [review]
Check slack variables

File: src/solver/optimisation/opt_appel_solveur_lineaire.cpp [hunk 3]
```diff
@@ -345,6 +352,17 @@ bool OPT_AppelDuSimplexe(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace, int NumIn
         logs.error() << "* Hydro reservoir impossible to manage with cumulative options \"hard "
                         "bounds without heuristic\"";
 
+        Optimization::InfeasibleProblemDiag diag(&Probleme, "(.)*");
```
Comment (guilpier-code) [inline src/solver/optimisation/opt_appel_solveur_lineaire.cpp]: we discussed it : **diagnostic** is not a good name for our analysis. Diag is worse.
Whatever name we choose, it should appear in full (not diag for diagnostic, and certainly not the short name for analysis)

Comment (guilpier-code) [reply src/solver/optimisation/opt_appel_solveur_lineaire.cpp]: Smothing else on this line : regular expression pattern should be an inner of **InfeasibleProblemDiag**

File: src/solver/infeasible-diag/problem.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+{
+    return mSolver->Solve();
+}
+
+InfeasibleProblemReport InfeasibleProblemDiag::produceReport()
+{
+    addSlackVariables();
+    if (mSlackVariables.size() == 0)
+        throw std::runtime_error(
+          "Cannot generate infeasibility report: no constraints have been pre-selected");
```
Comment (guilpier-code) [inline src/solver/infeasible-diag/problem.cpp]: For me, this message is unclear.
with the good regex pattern, is this possible that no slack variable was built ?
And what would it mean ?

File: src/solver/infeasible-diag/problem.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    buildObjective();
+    const MPSolver::ResultStatus s = Solve();
+    if (s != MPSolver::OPTIMAL && s != MPSolver::FEASIBLE)
+        throw std::runtime_error(
+          "Cannot generate infeasibility report: linear problem could not be solved");
```
Comment (guilpier-code) [inline src/solver/infeasible-diag/problem.cpp]: Would not this be more clear : "linear problem could not be solved, and infeasibility analysis could not help"

File: src/solver/infeasible-diag/report.cpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+{
+void InfeasibleProblemReport::append(const std::string& constraintName, double value)
+{
+    // TODO check prior existence of constraintName
+    mConstraints.push_back({constraintName, value});
+}
+
+void InfeasibleProblemReport::printLargest(unsigned int n)
+{
+    std::sort(std::begin(mConstraints), std::end(mConstraints), compareSlackSolutions);
```
Comment (guilpier-code) [inline src/solver/infeasible-diag/report.cpp]: Maybe missing some introductory message : "Infeasibility analysis gives following results :"

File: src/solver/infeasible-diag/report.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+{
+namespace Optimization
+{
+void InfeasibleProblemReport::append(const std::string& constraintName, double value)
+{
+    // TODO check prior existence of constraintName
+    mConstraints.push_back({constraintName, value});
+}
+
+void InfeasibleProblemReport::printLargest(unsigned int n)
```
Comment (guilpier-code) [inline src/solver/infeasible-diag/report.cpp]: **printLargest** : why not make this function's parameter a **InfeasibleProblemReport** member or constant ?
And in that case, we could call this method **print**

File: src/solver/cmake/infeasible-diag.cmake [hunk 8]
```diff
@@ -0,0 +1,12 @@
+set(SRC_INFEASIBLE_DIAG
```
Comment (guilpier-code) [inline src/solver/cmake/infeasible-diag.cmake]: Of course, if we change infeasible into something else, like analysis for instance, we should also change the name of the directory containing these files

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/infeasible-diag/problem.cpp (hunk 4 above)
Comment (flomnes) [reply src/solver/infeasible-diag/problem.cpp]: With the right regex, we will match at least 1 constraint (hydro levels for example), even if the user has not defined any binding constraints.

This check is added for extra safety

File: src/solver/infeasible-diag/report.cpp (hunk 7 above)
Comment (flomnes) [reply src/solver/infeasible-diag/report.cpp]: I don't understand this remark, can you elaborate ?

File: src/solver/optimisation/opt_appel_solveur_lineaire.cpp (hunk 3 above)
Comment (flomnes) [reply src/solver/optimisation/opt_appel_solveur_lineaire.cpp]: Name: I agree and will rename this class to **InfeasibleProblemAnalysis**.

I don't understand the 2nd remark.

File: src/solver/infeasible-diag/report.cpp (hunk 6 above)
Comment (flomnes) [reply src/solver/infeasible-diag/report.cpp]: done

File: src/solver/cmake/infeasible-diag.cmake (hunk 8 above)
Comment (flomnes) [reply src/solver/cmake/infeasible-diag.cmake]: ok

File: src/solver/optimisation/opt_appel_solveur_lineaire.cpp (hunk 3 above)
Comment (guilpier-code) [reply src/solver/optimisation/opt_appel_solveur_lineaire.cpp]: Sorry, let me be more clear : you pass the regex pattern **"(.)*"** to the constructor of **InfeasibleProblemAnalysis**.
This pattern should be a data of the class, and not be passed to this constructor, whatever this pattern is.

File: src/solver/infeasible-diag/report.cpp (hunk 7 above)
Comment (guilpier-code) [reply src/solver/infeasible-diag/report.cpp]: Sure : somewhere else in the code, you call : 
`printLargest(10);`
but the **n** (that is **10** here) that you pass, should it not be a inner data of the class **InfeasibleProblemReport** ?
I mean  : does the caller of **printLargest** have to known  that the function prints up to 10 incorrect constraints ?
And if you make this **n** an inner data, **printLargest** should be renamed **print()**

File: src/solver/infeasible-diag/problem.cpp (hunk 4 above)
Comment (guilpier-code) [reply src/solver/infeasible-diag/problem.cpp]: ok, that said, user will probably not know what is a **pre-selected** constraint.
So the message will be obscure to him.
Maybe the message could be : "Cannot generate infeasibility report: no incorrect constraint found"

File: src/solver/infeasible-diag/report.cpp (hunk 7 above)
Comment (flomnes) [reply src/solver/infeasible-diag/report.cpp]: OK, understood & accepted

File: src/solver/infeasible-diag/problem.cpp (hunk 4 above)
Comment (flomnes) [reply src/solver/infeasible-diag/problem.cpp]: I suggest: "no constraint have been selected". The message you suggest is misleading, it implies that the problem is feasible, even if it isn't.

File: src/solver/cmake/infeasible-diag.cmake (hunk 2 above)
Comment (flomnes) [reply src/solver/cmake/infeasible-diag.cmake]: done

File: src/solver/optimisation/opt_appel_solveur_lineaire.cpp (hunk 3 above)
Comment (flomnes) [reply src/solver/optimisation/opt_appel_solveur_lineaire.cpp]: Why not provide it to the constructor ?

**Comment by flomnes** [comment]
Contained in https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/431

**Pull request closed without merging**